samples/cpp/tutorial_code/videoio/astra/Astra.cpp

```diff
@@ -0,0 +1,158 @@
+#include <opencv2/videoio/videoio.hpp>
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+
+#include <iostream>
+#include <thread>
+#include <condition_variable>
+
+using namespace cv;
+using std::cout;
+using std::cerr;
+using std::endl;
+
+
+int main()
+{
+
+    //! [Open streams]
+    // Open color stream
+    VideoCapture colorStream(CAP_V4L2);
+    // Open depth stream
+    VideoCapture depthStream(CAP_OPENNI2_ASTRA);
+    //! [Open streams]
+
+    // Check that stream has opened
+    if (!colorStream.isOpened())
+    {
+        cerr << "ERROR: Unable to open color stream" << endl;
+        return 1;
+    }
+
+    // Check that stream has opened
+    if (!depthStream.isOpened())
+    {
+        cerr << "ERROR: Unable to open depth stream" << endl;
+        return 1;
+    }
+
+    // Set color stream size equal to depth stream's ones
+    //colorStream.set(CAP_PROP_FRAME_WIDTH, depthStream.get(CAP_PROP_FRAME_WIDTH));
+    //colorStream.set(CAP_PROP_FRAME_HEIGHT, depthStream.get(CAP_PROP_FRAME_HEIGHT));
+    //! [Setup streams]
+    // Set color and depth stream parameters
+    colorStream.set(CAP_PROP_FRAME_WIDTH,  640);
+    colorStream.set(CAP_PROP_FRAME_HEIGHT, 480);
+    depthStream.set(CAP_PROP_FRAME_WIDTH,  640);
+    depthStream.set(CAP_PROP_FRAME_HEIGHT, 480);
+    depthStream.set(CAP_PROP_OPENNI2_MIRROR, 0);
+    //! [Setup streams]
+
+    // Print color stream parameters
+    cout << "Color stream: "
+         << colorStream.get(CAP_PROP_FRAME_WIDTH) << "x" << colorStream.get(CAP_PROP_FRAME_HEIGHT)
+         << " @" << colorStream.get(CAP_PROP_FPS) << " fps" << endl;
+
+    //! [Get properties]
+    // Print depth stream parameters
+    cout << "Depth stream: "
+         << depthStream.get(CAP_PROP_FRAME_WIDTH) << "x" << depthStream.get(CAP_PROP_FRAME_HEIGHT)
+         << " @" << depthStream.get(CAP_PROP_FPS) << " fps" << endl;
+    //! [Get properties]
+
+    //! [Read streams]
+    // Start two reading threads
+    std::mutex mtx;
+    std::condition_variable dataReady;
+    bool isFinish = false;
+
+    std::thread depthReader([&depthStream, &dataReady, &isFinish]{
+        while (!isFinish)
+        {
+            // Grab new frame
+            bool frameReady = depthStream.grab();
+            if (frameReady)
+                dataReady.notify_all();
+        }
+    });
+
+    std::thread colorReader([&colorStream, &dataReady, &isFinish]{
+        while (!isFinish)
+        {
+            // Grab new frame
+            bool frameReady = colorStream.grab();
+            if (frameReady)
+                dataReady.notify_all();
+        }
+    });
+    //! [Read streams]
+
+    int errCode = 0;
+    Mat colorFrame, irFrame, depthMap, disparityMap;
+    while (true)
+    {
+        std::unique_lock<std::mutex> lk(mtx);
+        dataReady.wait(lk);
```

Comment:
> It may also be unblocked spuriously.

https://en.cppreference.com/w/cpp/thread/condition_variable/wait

Need to add while loop with condition check.

Reply:
done